### ISSUE TITLE
Endianness of EXT types

### ISSUE THREAD

**Issue by realh** [neovim/neovim#6620]
The API docs say that Buffer, Window and Tabpage types can be interchanged with integers. My initial experiments discovered them encoded in msgpack EXT objects as single bytes, but can they ever be more than one byte, and if so, are there any guarantees about their endianness? And should they be treated as unsigned or signed?

**Comment by justinmk** [comment]
They definitely can be more than one byte. Per [msgpack spec](https://github.com/msgpack/msgpack/blob/master/spec.md), contents of extension field are a byte array. Integers in msgpack are big-endian.

**Comment by bfredl** [comment]
I don't think `msgpack_rpc.txt` ever says that the EXT inner data is an msgpack INT, so this is a documentation issue.

**Comment by realh** [comment]
The bit of documentation I was referring to is in [nvim's msgpack-rpc page](https://neovim.io/doc/user/msgpack_rpc.html), where it says:

> An API method expecting one of these types (Buffer, Window or Tabpage) may be passed an integer instead, although they are not interchangeable. For example, a Buffer may be passed as an integer, but not a Window or Tabpage.

I don't think that's a mistake in the docs, I interpreted it to mean that vim stores these values as integers internally, but when it sends a handle over msgpack it encodes it as an EXT, while clients can send either EXT or integers the other way and vim will accept either. With what I've done so far, using `nvim_list_bufs`, `nvim_buf_get_name` etc, this appears to be true. But I had to make assumptions about how to treat a sequence of n bytes as an n-word integer, because the mapping is internal to vim, not defined in msgpack spec.

I guess I'm not supposed to try to decode the EXTs at all, but it isn't clear what to do instead, if I want to pass a number instead of EXT to `nvim_buf_` functions. Maybe that's what `nvim_buf_get_number` is for? It seems to return values identical to my decoding-to-int of the EXT handles.

So, yes, this is a documentation issue, I think clarifying the above would be a big help. If I'm right about `nvim_buf_get_number` etc just above, would you like me to draft some changes to the docs?

**Comment by bfredl** [comment]
> I guess I'm not supposed to try to decode the EXTs at all

This used to be the case, but after #4934 the id in the EXT is equal to buffer numbers (and later window ids when we merged that). We decided to mark the EXT representation as final, so it is safe for clients to decode it, but we never went around to document it.

`nvim_buf_get_number` should thus be documented as deprecated as well, as it is no longer needed.
